driver: timer: power: npcx: Add system timer and power management support in npcx series.

MulinChao opened the pull request:
This PR is separated into three CLs.

This first CL introduces a kernel device driver implemented by the internal
64/32-bit timers in Nuvoton NPCX series. Via these two kinds of timer,
the driver provides a standard "system clock driver" interface. 

The second CL introduces power management driver that improves the efficiency
of ec operation by adjusting the chip’s power consumption to the level
of activity required by the application in npcx series.
And this CL implements one power state, PM_STATE_SUSPEND_TO_IDLE, with
two sub-states for Zephyr power management system.
Sub-state 0 - "Deep Sleep" mode with “Instant” wake-up if residency time
               is greater or equal to 1 ms
Sub-state 1 - "Deep Sleep" mode with "Standard" wake-up if residency time
               is greater or equal to 201 ms

The third CL introduces power management device support in uart device.

Athought the uart device clock will be stopped after ec entered sleep or
deep sleep state and restore its clock automatically (there is no need
to implement code for suspending and resuming devices manually.), 
we still need to check the current status of uart device when it 
wants to change its power state to LOW or SUSPEND power state. It is 
crucial to forbid ec enters sleep or deep sleep state when uart device is 
busy with transmitting data. Or we will observe broken characters on the 
uart console.

MulinChao committed:
driver: timer: npcx: add system kernel timer support.

This CL introduces a kernel device driver implemented by the internal
64/32-bit timers in Nuvoton NPCX series. Via these two kinds of timer,
the driver provides an standard "system clock driver" interface.

It includes:
 - A system timer based on an ITIM64 (Internal 64-bit timer) instance,
   clocked by APB2 which freq is CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC.
 - Its prescaler is set to 1 and provide the kernel cycles reading
   without handling overflow mechanism.
 - A event timer based on an ITIM32 (Internal 32-bit timer) instance,
   clocked by LCLK which frequency is 32KHz and still activated when ec
   entered "idle/deep idle" power state for better power consumption.
 - Its prescaler is set to 1 and provide timeout event mechansim.
 - Compensate system timer which clock is gating for better power
   consumption after ec left"idle/deep idle" power state.

This CL passed starve, timer_api, and timer_monotonic test suites.

Signed-off-by: Mulin Chao <mlchao@nuvoton.com>

MulinChao committed:
driver: soc: power: npcx: Add power managerment support.

This CL introduces power management driver that improves the efficiency
of ec operation by adjusting the chip’s power consumption to the level
of activity required by the application in npcx series.

The following list summarizes the main properties of the various chip
power states. Please refer the power.c file for more detail.

Main power states in npcx series include:
- Active: Core, RAM and modules operate at the clocks generated by PLL.
- Idle: Enter this state when the Core executes WFI or WFE instruction.
- Sleep: clock is stopped for most of modules but PLL is enabled.
- Deep Sleep: As Sleep mode but PLL is disabled.
- Standby: All power rails are turned off besides standby and battery
  power rails.

And this CL implements one power state, PM_STATE_SUSPEND_TO_IDLE, with
two sub-states for Zephyr power management system.
Sub-state 0 - "Deep Sleep" mode with “Instant” wake-up if residency
              time is greater or equal to 1 ms
Sub-state 1 - "Deep Sleep" mode with "Standard" wake-up if residency
              time is greater or equal to 201 ms

Signed-off-by: Mulin Chao <mlchao@nuvoton.com>

MulinChao committed:
driver: uart: npcx: Add pm_deivce support for uart driver.

Since the uart device clock will be stopped after ec entered sleep or
deep sleep state and restore its clock automatically, there is no need
to implement code for suspending and resuming devices manually.

The driver still needs to check the current status of uart device when
it wants to change its power state to LOW or SUSPEND power state. It is
crucial to forbid ec enters sleep or deep sleep state when uart device
is busy with transmitting data. Or we will observe broken characters on
the uart console.

Signed-off-by: Mulin Chao <mlchao@nuvoton.com>

MulinChao commented:
Digging the build error log further, I found pm_policy_low_power_devices() is not implemented in tests/kernel/profiling/profiling_api test suite. (This test suite enabled CONFIG_PM_POLICY_APP but has no default pm_policy_low_power_devices() function. I thought this function should be added in the test suite, not the pm driver itself.

sjg20 commented on drivers/timer/npcx_itim_timer.c:
```diff
@@ -0,0 +1,247 @@
+ * - Setup time-out event and handle z_clock_announce() in its ISR.
+ * - Its prescaler is set to 1 and provide timeout event mechanism.
+ * - Compensate system timer which clock is gating after ec left
+ *   "sleep/deep sleep" power state.
```
What are these last three bullet points about?

MulinChao replied:
Indeed, it looks ambiguous. I will modify them for better explanations.

MulinChao replied:
Done.

sjg20 commented on drivers/timer/npcx_itim_timer.c:
```diff
@@ -0,0 +1,247 @@
+	cnt64l  = NPCX_ITIM64_MAX_HALF_CNT - cnt64l + 1;
+
+	/* Return current value of 64-bit counter value of system timer */
+	return ((((uint64_t)cnt64h2) << 32) | cnt64l);
```
drop outer ()

(in fact none is needed)

MulinChao replied:
Done.

sjg20 commented on drivers/timer/npcx_itim_timer.c:
```diff
@@ -0,0 +1,247 @@
+/* ITIM local inline functions */
+static inline uint64_t npcx_itim_get_sys_cyc64(void)
+{
+	uint32_t cnt64h1, cnt64h2, cnt64l;
```
minor point, but for naming I would suggest

cnt64h

and cnt64h_check

then use cnt64h below the loop. That makes it clearer what you are doing

MulinChao replied:
Indeed, that's clearer. Thanks.

MulinChao replied:
Done.

sjg20 commented on drivers/timer/npcx_itim_timer.c:
```diff
@@ -0,0 +1,247 @@
+	 * round up to next tick boundary.
+	 */
+	if (ticks <= 0)
+		ticks = 1;
```
{} sadly to keep Zephyr happy

MulinChao replied:
Done.  😄

sjg20 commented on drivers/timer/npcx_itim_timer.c:
```diff
@@ -0,0 +1,247 @@
+
+	k_spin_unlock(&lock, key);
+
+	/* Retrun how many ticks elapsed since last z_clock_announce() call */
```
Return

MulinChao replied:
Done.

sjg20 commented on drivers/timer/npcx_itim_timer.c:
```diff
@@ -0,0 +1,247 @@
+
+	k_spin_unlock(&lock, key);
+
+	/* Retrun how many cycles since system kernel timer start counting */
```
Return

MulinChao replied:
Done.

sjg20 reviewed (changes requested):
again I prefer lower-case hex

MulinChao commented:
> again I prefer lower-case hex

Done.

yperess commented on drivers/clock_control/clock_control_npcx.c:
```diff
@@ -100,6 +100,37 @@ static int npcx_clock_control_get_subsys_rate(const struct device *dev,
+
+	inst_pmc->PMCSR = 0;
+}
+#endif
```
Could you add a closing comment to make sure that the if/endif are associated?

```
#endif /* CONFIG_PM */
```

MulinChao replied:
Done.

yperess commented on drivers/espi/host_subs_npcx.c:
```diff
@@ -1050,5 +1050,17 @@ int npcx_host_init_subs_core_domain(const struct device *host_bus_dev,
 	irq_enable(DT_INST_IRQ_BY_NAME(0, p80_fifo, irq));
 #endif
 
+#if defined(CONFIG_PM)
+	/*
+	 * Configure the host access wake-up event triggered from a host
+	 * transaction on eSPI/LPC bus. No need for callback function.
+	 */
+	npcx_miwu_interrupt_configure(&host_sub_cfg.host_acc_wui,
+			NPCX_MIWU_MODE_EDGE, NPCX_MIWU_TRIG_HIGH);
+
+	/* Enable irq of interrupt-input module */
+	npcx_miwu_irq_enable(&host_sub_cfg.host_acc_wui);
+#endif
```
This is more readable as:

```
if (IS_ENABLED(CONFIG_PM)) {
  ...
}
```

MulinChao replied:
Done.

yperess commented on drivers/serial/uart_npcx.c:
```diff
@@ -356,13 +360,77 @@ static int uart_npcx_init(const struct device *dev)
+#ifdef CONFIG_PM_DEVICE
+static inline bool uart_npcx_device_is_transmitting(const struct device *dev)
+{
+#if defined(CONFIG_UART_INTERRUPT_DRIVEN)
+	/* The transmitted transaction is completed? */
+	return !uart_npcx_irq_tx_complete(dev);
+#else
+	/* No need fot polling mode */
+	return 0;
+#endif
```
Same here, prefer

```
if (IS_ENABLED(CONFIG_UART_INTERRUPT_DRIVEN))
        return ...
else
        return 0;
```

MulinChao replied:
Done.

yperess commented on drivers/serial/uart_npcx.c:
```diff
@@ -356,13 +360,77 @@ static int uart_npcx_init(const struct device *dev)
+	}
+	return ret;
+}
+#endif
```
Please add a closing comment:

```
#endif /* CONFIG_PM_DEVICE */
```

MulinChao replied:
Done.

yperess commented on drivers/timer/npcx_itim_timer.c:
```diff
@@ -0,0 +1,319 @@
+	cnt64l = NPCX_ITIM64_MAX_HALF_CNT - cnt64l + 1;
+
+	/* Return current value of 64-bit counter value of system timer */
+#if defined(CONFIG_PM)
+	return ((((uint64_t)cnt64h) << 32) | cnt64l) + cyc_sys_compensated;
+#else
+	return (((uint64_t)cnt64h) << 32) | cnt64l;
+#endif
```
Here too, prefer `IS_ENABLED` for readability.

MulinChao replied:
Replacing this with IS_ENABLED() macro will cause a compiler error since 'cyc_sys_compensated' is undeclared if CONFIG_PM is not enabled. If I remove line 69 & 74, we need to move 'cyc_evt_enter_deep_idle' variable to line 309 in case 'unused variable' warning. Does it improve readability if I do so?

yperess replied:
I see, up to you, personally I prefer to set line 71 to:
```
__maybe_unused static uint64_t cyc_sys_compensated;
```

That avoids the warning. Though looking at the code further I think you can just always declare cyc_sys_compensated which will default to 0, then you don't even need this if statement here and you can just always `+ cyc_sys_compensated`. The compiler should have an easy enough time figuring out that `cyc_sys_compensated` is always 0 and replace it accordingly.

* I recognize that I'm always in favor of offloading work to the compiler/optimizer in favor of simpler code, if you disagree I'm not going to block on this, just throwing my 2 cents.

MulinChao replied:
_maybe_unused seems doesn't work on Zephyr. But I use "__unused" instead of it. I agree with you that " offloading work to the compiler/optimizer in favor of simpler code" improves readability. Thanks for your comment. I learn a lot.

MulinChao replied:
Done.

yperess commented on drivers/timer/npcx_itim_timer.c:
```diff
@@ -0,0 +1,319 @@
+	/* Return elapsed cycles of 32-bit counter of event timer  */
+	return cnt2;
+}
+#endif
```
Closing comment here too please.

MulinChao replied:
Done.

yperess commented on drivers/timer/npcx_itim_timer.c:
```diff
@@ -0,0 +1,319 @@
+{
+	return  cyc_sys_compensated / SYS_CYCLES_PER_TICK;
+}
+#endif
+#endif
```
Closing comments on these too please.

MulinChao replied:
Done.

yperess commented on soc/arm/nuvoton_npcx/npcx7/power.c:
```diff
@@ -0,0 +1,157 @@
+	:: "r" (CONFIG_SRAM_BASE_ADDRESS)); /* A valid addr used for delay */  \
+	})
+
+#if defined(CONFIG_SOC_POWER_MANAGEMENT_TRACE)
+uint32_t cnt_sleep0;
+uint32_t cnt_sleep1;
+#endif
```
If we just have these as static will they not be taken out by the optimizer anyway if `!IS_ENABLED(CONFIG_SOC_POWER_MANAGEMENT_TRACE)`? Putting these in will allow us to use the more readable `IS_ENABLED` further down.

MulinChao replied:
Good idea. I will replace them with IS_ENABLED() macro later.

MulinChao replied:
Done.

yperess commented on soc/arm/nuvoton_npcx/npcx7/power.c:
```diff
@@ -0,0 +1,157 @@
+		switch (info.substate_id) {
+		case 0:	/* Sub-state 0: Deep sleep with instant wake-up */
+			npcx_power_enter_system_sleep(1, 1);
+#if defined(CONFIG_SOC_POWER_MANAGEMENT_TRACE)
+			cnt_sleep0++;
+#endif
```
Referencing the comment above, if we always define these we can just use `if (IS_ENABLED(...` here.

MulinChao replied:
Done.

yperess commented on soc/arm/nuvoton_npcx/npcx7/power.c:
```diff
@@ -0,0 +1,157 @@
+			break;
+		case 1:	/* Sub-state 1: Deep sleep with standard wake-up */
+			npcx_power_enter_system_sleep(1, 0);
+#if defined(CONFIG_SOC_POWER_MANAGEMENT_TRACE)
+			cnt_sleep1++;
+#endif
```
Same here

MulinChao replied:
Done.

yperess commented on soc/arm/nuvoton_npcx/npcx7/power.c:
```diff
@@ -0,0 +1,157 @@
+		}
+	}
+
+#if defined(CONFIG_SOC_POWER_MANAGEMENT_TRACE)
+	LOG_DBG("sleep: %d, deep sleep: %d", cnt_sleep0, cnt_sleep1);
+	LOG_INF("total ticks in sleep: %lld", npcx_clock_get_sleep_ticks());
+#endif
```
Same here

MulinChao replied:
Done.

yperess reviewed (changes requested):
Overall it looks good, minor nits here and there.

keith-zephyr commented on soc/arm/nuvoton_npcx/common/reg/reg_def.h:
```diff
@@ -1134,4 +1134,43 @@ struct smb_fifo_reg {
+struct itim32_reg {
+	volatile uint8_t reserved1;
+	/* 0x001: Internal Timer Prescaler */
+	volatile uint8_t ITPRE;
+	volatile uint8_t reserved2[2];
+	/* 0x004: Internal Timer Control and Status */
+	volatile uint8_t ITCTS;
+	volatile uint8_t reserved3[3];
+	/* 0x008: Internal 32-Bit Timer Counter */
+	volatile uint32_t ITCNT32;
```
The NPCX7nxC datasheet shows all register names as ITPRE32, ITCTS32, and ITCNT32.  I'm not sure about other datasheets.  But I think it improves readability of the driver if the 32-bit timer always includes "32" on the register name and same with the 64-bit timer (add "64").

MulinChao replied:
The original thought is to keep the ITCNT register and ITCNT_XXX field as the same prefix no matter 32-bit or 64-bit timers. But from the improving readability point of view, I agree with you. I will change all register's names to meet our datasheet and modified the prefix of the field name from NPCX_ITCTS_ to NPCX_ITCTSXX_.

MulinChao replied:
Done.

keith-zephyr commented on soc/arm/nuvoton_npcx/common/reg/reg_def.h:
```diff
@@ -1134,4 +1134,43 @@ struct smb_fifo_reg {
+struct itim64_reg {
+	volatile uint8_t reserved1;
+	/* 0x001: Internal Timer Prescaler */
+	volatile uint8_t ITPRE;
+	volatile uint8_t reserved2[2];
+	/* 0x004: Internal Timer Control and Status */
+	volatile uint8_t ITCTS;
```
Add "64" to the register names.

MulinChao replied:
Done.

keith-zephyr commented on drivers/timer/npcx_itim_timer.c:
```diff
@@ -0,0 +1,250 @@
+	/* Disable event timer if needed and wait for it to take effect */
+	if (IS_BIT_SET(evt_tmr->ITCTS, NPCX_ITCTS_ITEN)) {
+		evt_tmr->ITCTS &= ~BIT(NPCX_ITCTS_ITEN);
+		while (IS_BIT_SET(evt_tmr->ITCTS, NPCX_ITCTS_ITEN))
+			;
```
Can you put a upper bound on waiting of ITEN to set/clear for all instances?

MulinChao replied:
Done. Please see below. Thanks.

keith-zephyr commented on drivers/timer/npcx_itim_timer.c:
```diff
@@ -0,0 +1,250 @@
+	 * interrupt/wake-up sources, and clear timeout status bit before
+	 * enabling it.
+	 */
+	evt_tmr->ITCTS = BIT(NPCX_ITCTS_CKSEL) | BIT(NPCX_ITCTS_TO_WUE)
+			 | BIT(NPCX_ITCTS_TO_IE) | BIT(NPCX_ITCTS_TO_STS);
```
The datasheet indicates to wait 92us after changing CLKSEL before enabling the module.  Probably a good idea to wait inline after updates to ITCTS.

MulinChao replied:
Indeed. I will add a delay later.

MulinChao replied:
Done.

keith-zephyr commented on drivers/timer/npcx_itim_timer.c:
```diff
@@ -0,0 +1,250 @@
+			;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+	uint32_t delta_ticks = (uint32_t)((npcx_itim_get_sys_cyc64() -
+				cyc_sys_announced) / SYS_CYCLES_PER_TICK);
+
+	/* Store announced cycles of system timer */
+	cyc_sys_announced = npcx_itim_get_sys_cyc64();
+	k_spin_unlock(&lock, key);
```
This could be guarded with IS_ENABLED(CONFIG_TICKLESS_KERNEL) since delta_ticks is unused otherwise.

MulinChao replied:
Done.

keith-zephyr commented on soc/arm/nuvoton_npcx/npcx7/power.c:
```diff
@@ -0,0 +1,157 @@
+	} else {
+		switch (info.substate_id) {
+		case 0:	/* Sub-state 0: Deep sleep with instant wake-up */
+			npcx_power_enter_system_sleep(1, 1);
```
If you change the parameters to npcx_power_enter_system_sleep() to enums, then this code becomes a lot more readable.  For example:
    npcx_power_enter_system_sleep(DEEP_SLEEP, INSTANT_WAKE_UP);
    npcx_power_enter_system_sleep(DEEP_SLEEP, STANDARD_WAKE_UP);

MulinChao replied:
Done.

keith-zephyr commented on drivers/serial/uart_npcx.c:
```diff
@@ -362,6 +366,71 @@ static int uart_npcx_init(const struct device *dev)
+	/* The transmitted transaction is completed? */
+	return !uart_npcx_irq_tx_complete(dev);
+#else
+	/* No need fot polling mode */
```
Typo: "No need for polling mode"

MulinChao replied:
Done.

keith-zephyr commented on drivers/serial/uart_npcx.c:
```diff
@@ -362,6 +366,71 @@ static int uart_npcx_init(const struct device *dev)
+		 * If uart device is busy with transmitting, the driver will
+		 * stay in while loop and wait for the transaction is completed.
+		 */
+		while (uart_npcx_device_is_transmitting(dev))
+			;
```
Please put an upper bound on this loop.

MulinChao replied:
Done. The same assumption as above.

MulinChao commented:
Hi @yperess and @keith-zephyr :
I have finished most of the requests and left few open items for further discussion. Thanks for the effort of reviewing this.

keith-zephyr commented on drivers/timer/npcx_itim_timer.c:
```diff
@@ -0,0 +1,260 @@
+	/* Disable event timer if needed and wait for it to take effect */
+	if (IS_BIT_SET(evt_tmr->ITCTS32, NPCX_ITCTSXX_ITEN)) {
+		evt_tmr->ITCTS32 &= ~BIT(NPCX_ITCTSXX_ITEN);
+		while (IS_BIT_SET(evt_tmr->ITCTS32, NPCX_ITCTSXX_ITEN))
+			;
```
I would still like an upper bound put on the while loops.

MulinChao replied:
Sure. I will submit the other PR for this topic later.

MulinChao replied:
Hi @keith-zephyr 
I abstract all infinite loops and put them in two inline functions, npcx_itim_evt_enable() and npcx_itim_evt_disable().
After discussing with my hardware colleague, we think it is hardware proven behavior. (ITEN will be high after one clock from writing it as 1. Clearing it will take effect immediately.) Since it only costs 30us at most, do you think we need a timeout mechanism? If so, I will add the other PR for this later.

keith-zephyr replied:
These particular timeouts would only occur if there is a failure in the silicon (which I agree is very unlikely).  But it's good to eliminate unbounded loops whenever possible.  Please add this with another PR, but it's not high priority.

MulinChao replied:
Sure. I will add the other PR for unbounded loops after this PR is merged.

keith-zephyr commented on drivers/timer/npcx_itim_timer.c:
```diff
@@ -0,0 +1,260 @@
+	} else {
+		/* Enable event timer for ticking and wait to it take effect */
+		evt_tmr->ITCTS32 |= BIT(NPCX_ITCTSXX_ITEN);
+		while (!IS_BIT_SET(evt_tmr->ITCTS32, NPCX_ITCTSXX_ITEN)) {
+			continue;
```
Same here - please prevent an infinite loop in case of a hardware failure.

MulinChao committed:
tests: kernel: profiling: Add customized pm device policy handler.

Besides implementing a customized pm_policy_next_state() in the
application layer, a customized device policy handler of power
management, pm_policy_low_power_devices(), is also needed if
CONFIG_PM_POLICY_APP is enabled. This CL adds this function to prevent
build errors.

Signed-off-by: Mulin Chao <mlchao@nuvoton.com>

MulinChao commented:
> Digging the build error log further, I found pm_policy_low_power_devices() is not implemented in tests/kernel/profiling/profiling_api test suite. (This test suite enabled CONFIG_PM_POLICY_APP but has no default pm_policy_low_power_devices() function. I thought this function should be added in the test suite, not the pm driver itself.

A new [CL](https://github.com/zephyrproject-rtos/zephyr/pull/31697/commits/0358aeed8aa65160e1c6c8367160a1c2d62f2a7e)  is added in this PR for the build issue of tests/kernel/profiling/profiling_api test suite in case buildkite/zephyr blocks this PR.

MulinChao commented:
Rebasing since [PR31578](https://github.com/zephyrproject-rtos/zephyr/pull/31578)
